### ISSUE TITLE
The implementation of pad bucket size causes the loss to be inconsistent with before

### ISSUE THREAD

**Issue by Richie-yan** [NVIDIA/Megatron-LM#657]
After with the pad bucket size code, I noticed a difference in the training loss compared to before. 
the code below caused a diff order in the opt params generated
<img width="793" alt="截屏2024-01-10 上午10 07 17" src="https://github.com/NVIDIA/Megatron-LM/assets/41471499/b6b8cc0a-43f9-44d0-8ea2-0cf8a5285d31">
<img width="639" alt="截屏2024-01-10 上午10 10 01" src="https://github.com/NVIDIA/Megatron-LM/assets/41471499/618e9147-f10f-401e-b9b9-1fc0f908f678">
The generated param_range_map, which is order-dependent, is used for sharding optimizer parameters. Therefore, when using the pad bucket code, the order of the resulting optimizer parameters differs from the previous one.

which in turn led to an inconsistency in the order of gradient norm accumulation when using clip grad. Consequently, the final loss differed from what it was when the pad bucket size code was not used.
@deepakn94 @jaredcasper

**Comment by deepakn94** [comment]
Yes, we were aware of this when making the change; it is unfortunate the losses no longer match exactly, but the convergence curves should be similar (i.e., the convergence rate is un-changed). Let me know if you are seeing something different.

**Comment by Richie-yan** [comment]
I haven't seen a change in the convergence rate for the time being, so for now, this loss difference is acceptable.
Or, can it be changed to be consistent with the previous order?

**Comment by deepakn94** [comment]
That unfortunately will be difficult and is not a priority right now.

**Comment by deepakn94** [comment]
I am going to close the issue, feel free to re-open if you run into any other issues.
